### PR TITLE
Remove the logic allowing lighthouse to update it's own ENR

### PR DESCRIPTION
**Pull request by pawanjay176**
## Issue Addressed

Addresses #644 

## Proposed Changes

For accommodating the `zero-ports` flag, we find free ports using the `unused_ports` function and explicitly set the `libp2p_port` and `discovery_port` instead of setting it to zero as before. This removes the need for lighthouse to update its own ENR as the ports are known beforehand.

Also returns an error if the `discovery-addr` 0.0.0.0 and not an explicit ip address. This is to ensure that nodes can always be discovered through their ENRs.

## Additional info

`unused_ports()` function is not ideal but should work well enough since zero-ports flag is for mainly used for spinning up local testnets.

### CODE REVIEW

File: beacon_node/src/config.rs [hunk 1]
```diff
@@ -254,14 +255,29 @@ pub fn get_configs<E: EthSpec>(
...
-        client_config.network.libp2p_port = 0;
-        client_config.network.discovery_port = 0;
+        if client_config.network.discovery_address == IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)) {
+            client_config.network.discovery_address = "127.0.0.1"
+                .parse()
+                .map_err(|e| format!("Invalid discovery address: {}", e))?
+        }
+        client_config.network.libp2p_port =
+            unused_port().map_err(|e| format!("Failed to get port for libp2p: {}", e))?;
+        client_config.network.discovery_port =
```
[the remark this answers is not in the document]
Comment (pawanjay176) [reply beacon_node/src/config.rs]: Yup, will fix

File: beacon_node/src/config.rs [hunk 2]
```diff
@@ -270,14 +271,29 @@ pub fn get_configs<E: EthSpec>(
...
+        }
+        client_config.network.libp2p_port =
+            unused_port("tcp").map_err(|e| format!("Failed to get port for libp2p: {}", e))?;
+        client_config.network.discovery_port =
+            unused_port("udp").map_err(|e| format!("Failed to get port for discovery: {}", e))?;
         client_config.rest_api.port = 0;
         client_config.websocket_server.port = 0;
     }
 
+    // ENR ip needs to be explicit for node to be discoverable
```
Comment (pawanjay176) [inline beacon_node/src/config.rs]: @AgeManning This check is restrictive as we have to explicitly pass the `discovery-address` even for nodes which aren't bootnodes. 

Non-bootnodes will work without this check as their ENR will get updated after dialing the bootnode. 
I can't think of a good way to add this check just for bootnodes.

Comment (AgeManning) [reply beacon_node/src/config.rs]: I think we can just set it to "127.0.0.1" as you've done before. But perhaps log a warning saying that we have done this.

File: beacon_node/src/config.rs [hunk 3]
```diff
@@ -270,14 +271,29 @@ pub fn get_configs<E: EthSpec>(
...
+     * from lighthouse.
+     * Discovery address is set to localhost by default.
      */
     if cli_args.is_present("zero-ports") {
-        client_config.network.libp2p_port = 0;
-        client_config.network.discovery_port = 0;
+        if client_config.network.discovery_address == IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)) {
+            client_config.network.discovery_address = "127.0.0.1"
+                .parse()
+                .map_err(|e| format!("Invalid discovery address: {}", e))?
```
Comment (AgeManning) [inline beacon_node/src/config.rs]: This can just be an `expect("Valid IP address")`.

**Comment by pawanjay176** [comment]
@AgeManning seems like we can't get away with lighthouse not updating its own ENR with the addition of attestation subnet bitfield in the ENR.

**Comment by AgeManning** [comment]
@pawanjay176 - Interesting point about the bitfield. 

I think we will have to add that functionality then. I'll make an issue

**Pull request merged**